File: sui/ts/src/wormhole.ts
import type { Transaction, TransactionArgument } from '@mysten/sui/transactions';
import type { SharedObjectInfo } from './utils.ts';
import { addSharedObject } from './utils.ts';

export function verifyVaa(
  tx: Transaction,
  req: {
    wormholePackage: string
    wormholeState: SharedObjectInfo
    clock: TransactionArgument
    vaa: Uint8Array
  },
): TransactionArgument {
  const wormholeStateArg = addSharedObject(tx, req.wormholeState, false);
  const vaaArg = tx.pure.vector('u8', req.vaa);

  return tx.moveCall({
    package: req.wormholePackage,
    module: 'vaa',
    function: 'parse_and_verify',
    typeArguments: [],
    arguments: [wormholeStateArg, vaaArg, req.clock],
  });
}

File: sui/rust/src/wormhole.rs
use crate::utils::{add_bytes_to_tx, add_shared_object};
use sui_sdk_types::Argument;
use sui_transaction_builder::{Function, TransactionBuilder};

use crate::pyth::UpdatePriceFeeds;
use anyhow::Result;

/// Returns VAA
pub(crate) fn verify_vaas(tx: &mut TransactionBuilder, req: &UpdatePriceFeeds) -> Result<Argument> {
    let wormhole_state_arg = add_shared_object(tx, req.wormhole_state, false)?;

    let vaa_arg = add_bytes_to_tx(tx, req.vaa);

    let function = Function::new(
        req.wormhole_package,
        "vaa".parse()?,
        "parse_and_verify".parse()?,
        vec![],
    );

    let result = tx.move_call(function, vec![wormhole_state_arg, vaa_arg, req.clock]);

    Ok(result)
}
